### ISSUE TITLE
Issue#1: 多岐選択肢 Instruction Finetuning (JCommonsenseQA, unsloth)

### ISSUE THREAD

**Issue by Ishihara-Masabumi** [Ishihara-Masabumi/llama32_unsloth_sft_dpo#1]
# Issue #1: 多岐選択肢 Instruction Finetuning (unsloth 版)

## 目的
元実装(職務経歴書記載)と同条件で、JCommonsenseQA を用いた多岐選択肢SFTを unsloth ベースで再現し、精度を比較する。

## 環境構築

| 項目 | 値 |
| --- | --- |
| GPU | NVIDIA L40S 46GB |
| OS | Ubuntu (Linux 6.8.0-1045-aws) |
| Python | 3.11.15 |
| PyTorch | 2.5.0+cu124 |
| unsloth | 2026.5.2 |
| transformers | 5.5.0 |
| trl | 0.24.0 |
| peft | 0.19.1 |
| bitsandbytes | 0.49.2 |
| conda env | `llama32_unsloth` |

セットアップ:

```bash
bash setup_env.sh
# 既知 issue: 最新 torchao が torch.int1 を要求しエラーになるので削除
conda activate llama32_unsloth && pip uninstall -y torchao
```

## データセット

- `sbintuitions/JCommonsenseQA`
- train: **8,939** 件 (※職務経歴書では 9,759 件と記載。HFのsbintuitions版splitが正規化された結果と思われる)
- validation: **1,119** 件 (職務経歴書と一致)
- 5択 (A〜E)

## ハイパーパラメータ (元実装に揃える)

| 項目 | 値 |
| --- | --- |
| ベースモデル | `meta-llama/Llama-3.2-3B-Instruct` |
| 量子化 | 4bit (bitsandbytes) |
| LoRA R | 16 |
| LoRA alpha | 32 |
| LoRA dropout | 0.05 |
| target modules | q,k,v,o,gate,up,down_proj |
| Optimizer | AdamW (8bit) |
| LR | 2e-4, cosine, warmup 3% |
| Batch | per_device=4, grad_accum=4 → effective 16 |
| Epochs | 3 |
| max_seq_len | 1024 |
| Chat template | llama-3.1 |
| seed | 42 |

## フォーマット2系統

**letter_text**: assistant が `A. <選択肢テキスト>` で答える

**answer_tag**: assistant が `[ANSWER] A` で答える

## コマンドライン

```bash
# 1) データ整形
python scripts/prepare_jcommonsenseqa.py --out-dir data/jcommonsenseqa

# 2) ベースライン評価
python scripts/eval_jcommonsenseqa.py \
    --model meta-llama/Llama-3.2-3B-Instruct \
    --output reports/issue1_eval_baseline.json

# 3) SFT (letter_text)
python scripts/train_mcq_sft_unsloth.py \
    --train-jsonl data/jcommonsenseqa/train.letter_text.jsonl \
    --output-dir outputs/mcq_sft_letter_text

# 4) 評価 (letter_text)
python scripts/eval_jcommonsenseqa.py \
    --adapter outputs/mcq_sft_letter_text \
    --output reports/issue1_eval_letter_text.json

# 5) SFT (answer_tag)
python scripts/train_mcq_sft_unsloth.py \
    --train-jsonl data/jcommonsenseqa/train.answer_tag.jsonl \
    --output-dir outputs/mcq_sft_answer_tag

# 6) 評価 (answer_tag)
python scripts/eval_jcommonsenseqa.py \
    --adapter outputs/mcq_sft_answer_tag \
    --output reports/issue1_eval_answer_tag.json
```

`run_issue1_mcq_sft.sh` に上記を一括で実行する形でまとめてある。

## 評価結果 (実行後に追記)

| モデル | Acc (validation 1,119) | unparsable |
| --- | --- | --- |
| baseline (Llama-3.2-3B-Instruct) | _TBD_ | _TBD_ |
| SFT (letter_text "X. 回答") | _TBD_ | _TBD_ |
| SFT (answer_tag "[ANSWER] X") | _TBD_ | _TBD_ |

## 元実装(職務経歴書)との比較

| モデル | 元実装 | unsloth 版 | 差分 |
| --- | --- | --- | --- |
| baseline | 22.16% | _TBD_ | _TBD_ |
| letter_text | 85.08% | _TBD_ | _TBD_ |
| answer_tag | 85.43% | _TBD_ | _TBD_ |

## メモ

- 元実装で baseline=22.16% という低い数値は、評価パーサが厳格 (おそらく `X.` または `[ANSWER] X` のみを許容) で、未学習のベースモデルが期待フォーマットを出さないために unparsable→不正解 となるためと推測。
- 本リポジトリの評価パーサは `[ANSWER] X` > `^X.` > `\bX\b` の順で寛容。スモークテスト 5件で baseline=80% と高めに出るため、職務経歴書の数値と直接比較する際は注意。

**Comment by Ishihara-Masabumi** [comment]
# Issue #1 追加検証: 厳密パーサによる再評価

## 厳密パーサの定義

| モード | 受理する出力 |
| --- | --- |
| `lenient` (これまで採用) | `[ANSWER] X` → 行頭 `X.` → 単独 `\bX\b` の順で寛容に拾う |
| `answer_tag` | `[ANSWER] X` のみ受理。それ以外は unparsable→不正解扱い |
| `letter_text` | 行頭/先頭 `X. ...` のみ受理。それ以外は unparsable→不正解扱い |

## 結果一覧

| モデル | パーサ | Acc | Correct | Unparsable |
| --- | --- | --- | --- | --- |
| baseline | lenient | 72.48% | 811/1119 | 35 |
| baseline | strict `answer_tag` | **0.00%** | 0/1119 | 1119 |
| baseline | strict `letter_text` | **0.80%** | 9/1119 | 1090 |
| SFT (letter_text) | lenient | 81.95% | 917/1119 | 0 |
| SFT (letter_text) | strict `letter_text` | **81.95%** | 917/1119 | 0 |
| SFT (answer_tag) | lenient | 82.31% | 921/1119 | 0 |
| SFT (answer_tag) | strict `answer_tag` | **82.31%** | 921/1119 | 0 |

## 考察

### 1. SFT 後はパーサ非依存

SFT を経たモデルは、訓練した形式(`X.` または `[ANSWER] X`)で**100%出力する**。厳密パーサでも寛容パーサでも全く同じ結果(917 / 921)に着地し、unparsable は 0 件。つまり SFT 後の精度数値は評価方法に依存しない。

### 2. baseline の 22.16% は本構成では再現できない

職務経歴書の baseline **22.16%** は、

- 厳密 `answer_tag` (0.00%) でも
- 厳密 `letter_text` (0.80%) でも
- 寛容 (72.48%) でも

いずれにも一致しなかった。したがって元実装は本リポと違う評価方式と推測される。可能性:
1. **両形式併用 + 追加正規表現**: `[ANSWER] X` と `X.` の両方を許容するさらに別パターン(例: 文末 `... X` や `回答は X`)を持つパーサ
2. **対数尤度ランキング** (lm-eval-harness 形式): 5択を生成させず、各選択肢の条件付き対数尤度を比較してアーグマックスを取る方式。Llama-3.2-3B は何ら追加学習なしで MCQ を 22% 程度(チャンス +1%) で当てる傾向と整合
3. **強い拒否系プロンプト**: モデルに「A/B/C/D/E のいずれか1文字だけ返す」と強要するゼロショットプロンプト

特に (2) の対数尤度ランキングが、5択チャンスレート 20% に近い 22% という数値とよく合うため最有力。

### 3. unsloth 実装の評価としては SFT 後 81.95% / 82.31% が妥当

evaluation parser 依存性が無いため、後段の Issue#2/#3 でも本値を unsloth 実装の MCQ-SFT 基準点として使える。

## 元実装(職務経歴書)との比較(更新版)

| モデル | 元実装 | unsloth 実装(本リポの厳密パーサ) | 差分 |
| --- | --- | --- | --- |
| baseline | 22.16% | 0.00%〜72.48% (定義依存) | — 直接比較不可 |
| letter_text | 85.08% | 81.95% | -3.13 pt |
| answer_tag | 85.43% | 82.31% | -3.12 pt |

SFT 後の精度差は約 -3pt で安定しており、本差分は以下が要因と推定(順):
1. 学習データ件数差(8,939 vs 9,759)
2. unsloth のプリ量子化モデル (`unsloth/llama-3.2-3b-instruct-unsloth-bnb-4bit`) を使用
3. シード/乱数性(1run のみ実施)
